### PR TITLE
Implement  #433

### PR DESCRIPTION
**Pull request by mcarton**
This is a WIP, I still have to figure out how to handle this:
```rust
fn insert_other_if_absent<K: Eq + Hash, V>(m: &mut HashMap<K, V>, k: K, o: K, v: V) {
    if !m.contains_key(&k) { m.insert(o, v); }
    //                  ^             ^
}
```
and probably lots of other potential problems.

### CODE REVIEW

**Comment by Manishearth** [comment]
This fails the dogfood test btw https://travis-ci.org/Manishearth/rust-clippy/builds/98226307#L236

File: src/hashmap.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+declare_lint! {
+    pub HASHMAP_ENTRY,
+    Warn,
+    "use of `contains_key` followed by `insert` on a `HashMap`"
+}
+
+pub const HASHMAP_PATH: [&'static str; 5] = ["std", "collections", "hash", "map", "HashMap"];
+
+#[derive(Copy,Clone)]
+pub struct HashMap;
```
Comment (Manishearth) [inline src/hashmap.rs]: Probably should be called HashMapLint or something to avoid confusion

**Comment by mcarton** [comment]
I completely forgot about that one.
Can be merged now.

**Comment by Manishearth** [comment]
needs a rebase

File: src/hashmap.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+///
+/// **Why is this bad?** Using `HashMap::entry` is more efficient.
+///
+/// **Known problems:** Some false negatives, eg.:
+/// ```
+/// let k = &key;
+/// if !m.contains_key(k) { m.insert(k.clone(), v); }
+/// ```
+///
+/// **Example:** `if !m.contains_key(&k) { m.insert(k, v) }`
```
Comment (Manishearth) [inline src/hashmap.rs]: I'd like an example of how it's fixed

File: src/hashmap.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    if_let_chain! {
+        [
+            let ExprMethodCall(ref name, _, ref params) = expr.node,
+            params.len() == 3,
+            name.node.as_str() == "insert",
+            get_item_name(cx, map) == get_item_name(cx, &*params[0]),
+            is_exp_equal(cx, key, &params[1])
+        ], {
+            span_help_and_lint(cx, HASHMAP_ENTRY, span,
+                               "usage of `contains_key` followed by `insert` on `HashMap`",
```
Comment (Manishearth) [inline src/hashmap.rs]: it would be nice if this suggested how to fix the entire if block

**Comment by mcarton** [comment]
Rebased and added better help message.

**Comment by Manishearth** [comment]
Awesome, thanks!

**Pull request merged**